### DISCUSSION TITLE
How to login in request specs without sending requests

### DISCUSSION THREAD

**Discussion by Samuelodan** [janko/rodauth-rails#243] [Q&A] [answered]
Hi there,
I want to know if it's possible to do this. I thought of logging in the user with an internal request, but I don't know how to retrieve the user's JWT after that since there's no HTTP request.
`request.env.fetch("rodauth.v1").session_jwt` (for V1::RodauthMain) seems like it wouldn't work either.

Any help will be appreciated. Thanks.

**Comment by janko** [comment]
Why do you want to login without making a login request?

**Reply by Samuelodan** [reply]
I think it might cut the overall time for the test suite, though, I don't know if it'll even be significant.

**Reply by janko** [reply, marked as answer]
Rodauth doesn't offer way to mocking login session in tests, similar to `Clearance::Backdoor` or Warden's `on_next_request`. In system tests I can imagine some overhead, though I would like to get some numbers from a real world app on what percentage of the test suite runtime is spent logging in, to motivate adding such functionality.

In requests specs I can imagine minimal overhead from actually hitting the login endpoint. Again, if you have some numbers on what amount of time goes to login compared to rest of the tests, that would be useful.

**Reply by Samuelodan** [reply]
Okay, thank you, Janko. There's no real overhead yet as the app is still in the early stages. I guess I was just curious to know if I could do it some other way.
